### PR TITLE
Esdt attribute annotation for type export in abi 

### PR DESCRIPTION
**Pull request by mihaicalinluca**
Annotating contracts/modules like this:

```rust
#[multiversx_sc::contract]
#[esdt_attribute("TICKER1", BigUint)]
#[esdt_attribute("TICKER2", ManagedBuffer)]
#[esdt_attribute("TICKER3", u32)]
#[esdt_attribute("STRUCT1", AbiEnum)]
#[esdt_attribute("STRUCT2", AbiManagedType<Self::Api>)]
#[esdt_attribute("OnlyInEsdt", OnlyShowsUpInEsdtAttr)]
pub trait AbiTester {
```

... yields ESDT attribute ABI files like this (`OnlyInEsdt.esdt-abi.json`):

```json
{
    "esdtAttribute": {
        "ticker": "OnlyInEsdt",
        "type": "OnlyShowsUpInEsdtAttr"
    },
    "types": {
        "OnlyShowsUpAsNested10": {
            "type": "struct",
            "docs": [
                "Tests that the ABI generator also fetches types that only appear as fields."
            ]
        },
        "OnlyShowsUpInEsdtAttr": {
            "type": "struct",
            "fields": [
                {
                    "name": "field",
                    "type": "OnlyShowsUpAsNested10"
                }
            ]
        }
    }
}
```

### CODE REVIEW

File: contracts/feature-tests/abi-tester/abi_tester_expected_main.abi.json [hunk 1]
```diff
@@ -790,5 +790,6 @@
                 }
             ]
         }
-    }
+    },
+    "esdtAttributes": []
```
Comment (andrei-marinica) [inline contracts/feature-tests/abi-tester/abi_tester_expected_main.abi.json:794]: The abi-tester is a great place to add a few tests of the new feature.

File: framework/base/src/abi/contract_abi.rs [hunk 2]
```diff
@@ -1,6 +1,29 @@
 use super::*;
 use alloc::{string::String, vec::Vec};
 
+// #[derive(Clone, Debug)]
```
Comment (andrei-marinica) [inline framework/base/src/abi/contract_abi.rs:4]: Delete commented code.

File: framework/derive/src/parse/attributes/trait_prop_names.rs [hunk 3]
```diff
@@ -1,3 +1,4 @@
 pub(super) static PROP_ONLY_OWNER: &str = "only_owner";
 pub(super) static PROP_ADMIN_OWNER: &str = "only_admin";
 pub(super) static PROP_ALLOW_MULTIPLE_VAR_ARGS: &str = "allow_multiple_var_args";
+pub(super) static PROP_ESDT_ATTRIBUTE: &str = "esdt_attribute";
```
Comment (andrei-marinica) [inline framework/derive/src/parse/attributes/trait_prop_names.rs:4]: Call a `cargo fmt` to sort out missing endlines at the end of files, and more.

File: framework/derive/src/parse/attributes/util.rs [hunk 4]
```diff
@@ -12,6 +15,66 @@ pub(super) fn is_attribute_with_no_args(attr: &syn::Attribute, name: &str) -> bo
     false
 }
 
+pub(super) fn is_attribute_with_one_type_arg(
+    attr: &syn::Attribute,
+    name: &str,
+) -> (bool, Option<EsdtAttribute>) {
```
Comment (andrei-marinica) [inline framework/derive/src/parse/attributes/util.rs:21]: Looks like the bool is redundant, the Option is enough.

File: framework/derive/src/parse/attributes/util.rs [hunk 5]
```diff
@@ -12,6 +15,66 @@ pub(super) fn is_attribute_with_no_args(attr: &syn::Attribute, name: &str) -> bo
     false
 }
 
+pub(super) fn is_attribute_with_one_type_arg(
```
Comment (andrei-marinica) [inline framework/derive/src/parse/attributes/util.rs:18]: prefix `is_` is only appropriate when the result is just a bool

File: framework/derive/src/parse/trait_argument_parse.rs [hunk 6]
```diff
@@ -49,3 +50,14 @@ fn process_allow_multiple_var_args_argument(
     }
     has_attr
 }
+
+fn process_esdt_attribute_argument(
+    attr: &syn::Attribute,
+    arg_metadata: &mut TraitProperties,
+) -> bool {
+    let (has_attr, esdt_attribute) = is_esdt_attribute_prop(attr);
+    if has_attr {
```
Comment (andrei-marinica) [inline framework/derive/src/parse/trait_argument_parse.rs:59]: Again, the bool is redundant.
if let Some(esdt_attribute) = ...

File: framework/meta/Cargo.toml [hunk 7]
```diff
@@ -25,6 +25,7 @@ template-test-current = []
 template-test-released = []
 
 [dependencies]
+proc-macro2 = "1.0.66"
```
Comment (andrei-marinica) [inline framework/meta/Cargo.toml:28]: Only the derive crate knows about proc-macro2

**Review by andrei-marinica — CHANGES_REQUESTED** [review]

File: contracts/feature-tests/abi-tester/src/abi_tester.rs [hunk 8]
```diff
@@ -16,6 +16,9 @@ use only_nested::*;
 /// Note: any change in this contract must also be reflected in `abi_test_expected.abi.json`,
 /// including Rust docs.
 #[multiversx_sc::contract]
+#[esdt_attribute("TICKER-1", BigUint)]
+#[esdt_attribute("TICKER-2", ManagedBuffer)]
+#[esdt_attribute("TICKER-3", u32)]
```
Comment (andrei-marinica) [inline contracts/feature-tests/abi-tester/src/abi_tester.rs:21]: Two more things: 
- also add a ticker with a structure/enum. You already have `AbiManagedType` and `AbiEnum` in the contract 
- create a structure `OnlyShowsUpInEsdtAttributes`, doesn't need to contain any fields, the point of it is, we want to see if all structures in esdt attributes get added to the types section of the ABI correctly.

File: framework/base/src/abi/contract_abi.rs [hunk 9]
```diff
@@ -1,6 +1,21 @@
...
 use alloc::{string::String, vec::Vec};
 
+#[derive(Clone, Debug)]
+pub struct EsdtAttributeAbi {
+    pub ticker: &'static str,
+    pub ty: TypeName,
+}
+
+impl EsdtAttributeAbi {
+    pub fn get_type<T: TypeAbi>(arg_name: &'static str) -> EsdtAttributeAbi {
```
Comment (andrei-marinica) [inline framework/base/src/abi/contract_abi.rs:11]: Rename to `new`

File: framework/derive/src/generate/abi_gen.rs [hunk 10]
```diff
@@ -196,6 +196,21 @@ fn generate_supertrait_snippets(contract: &ContractTrait) -> Vec<proc_macro2::To
...
 }
 
+fn generate_esdt_attribute_snippets(contract: &ContractTrait) -> Vec<proc_macro2::TokenStream> {
+    contract
+        .trait_attributes
+        .esdt_attribute
+        .iter()
+        .map(|esdt_attr| {
+            let attr = esdt_attr.clone();
+            let (ticker, ty) = attr.into_tuple();
```
Comment (andrei-marinica) [inline framework/derive/src/generate/abi_gen.rs:206]: Looks a little over-engineered.
```
let ticker = esdt_attr.ticker;
let ty = &esdt_attr.ty;
```
No need for clone or into_tuple (at all).

File: framework/derive/src/parse/attributes/util.rs [hunk 11]
```diff
@@ -12,6 +15,66 @@ pub(super) fn is_attribute_with_no_args(attr: &syn::Attribute, name: &str) -> bo
...
+
+            let first_literal = match iter.next() {
+                Some(TokenTree::Literal(literal)) => {
+                    let literal_string = literal.to_string();
+                    literal_string.trim_matches('\"').to_string()
+                },
+                _ => panic!("Expected a literal as the first token in the attribute argument"),
+            };
+
+            let symbol: &'static str = Box::leak(Box::new(first_literal));
```
Comment (andrei-marinica) [inline framework/derive/src/parse/attributes/util.rs:41]: Memory leak!
No need for static str in derive. Use String! The static str is a trick we employ when generating code (might remove it there too...). There's no need to have it here.

File: framework/derive/src/parse/attributes/util.rs [hunk 12]
```diff
@@ -12,6 +15,66 @@ pub(super) fn is_attribute_with_no_args(attr: &syn::Attribute, name: &str) -> bo
...
 
+pub(super) fn get_attribute_with_one_type_arg(
+    attr: &syn::Attribute,
+    name: &str,
+) -> Option<EsdtAttribute> {
+    let attr_path = &attr.path;
+    if let Some(first_seg) = attr_path.segments.first() {
+        if first_seg.ident == name {
+            let tokens = attr.tokens.clone();
+            let group = match syn::parse2::<Group>(tokens) {
```
Comment (andrei-marinica) [inline framework/derive/src/parse/attributes/util.rs:26]: Do we need to _parse_ the tokens?
I would expect an `if let Group(g) = tokens ... ` should be enough.

**Review by andrei-marinica — CHANGES_REQUESTED** [review]

**Review by andrei-marinica — APPROVED** [review]

**Review by alyn509 — APPROVED** [review]

**Comment by razvanphp** [comment]
So we still need to add the annotations for this to happen? `#[derive(TypeAbi)]` is not enough?

**Pull request merged**